include/boost/windows_sspi/stream.hpp

```diff
@@ -68,61 +68,43 @@ template <typename NextLayer, typename ConstBufferSequence> struct async_write_i
 
 // TODO: Move away from this file
 template <typename NextLayer, typename MutableBufferSequence> struct async_read_impl : boost::asio::coroutine {
-  async_read_impl(NextLayer& next_layer, const MutableBufferSequence& buffer, std::shared_ptr<detail::sspi_impl> sspi_impl)
+  async_read_impl(NextLayer& next_layer, const MutableBufferSequence& buffers, std::shared_ptr<detail::sspi_impl> sspi_impl)
     : m_next_layer(next_layer)
-    , m_buffer(buffer)
+    , m_buffers(buffers)
     , m_sspi_impl(std::move(sspi_impl)) {
   }
 
   template <typename Self> void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
     BOOST_ASIO_CORO_REENTER(*this) {
-      while(m_sspi_impl->decrypt.decrypted_data.empty()) {
-        // TODO: Find some way to make the sspi_impl, the decrypt
-        // function or something else be responsible for keeping track
-        // of state and buffer(s)
-        // TODO: Fix so overflow cannot happen (we don't need to read
-        // more data unless DecryptMessage asks us to).
-        BOOST_ASSERT(m_sspi_impl->decrypt.encrypted_data.size() < 0x10000);
-        m_message.resize(0x10000 - m_sspi_impl->decrypt.encrypted_data.size());
+      while(m_sspi_impl->decrypt() == detail::sspi_decrypt::state::data_needed) {
+        // TODO: Use a fixed size buffer instead
+        m_input.resize(0x10000);
         BOOST_ASIO_CORO_YIELD net::async_read(m_next_layer,
-                                              net::buffer(m_message),
+                                              net::buffer(m_input),
```

Comment:
async_read will 'hang' until 64k bytes read or end of stream. Is this the intention? The inititation function is async_read_some, which a user would expect to complete quickly if there was already any data in the tcp receive window.

You may want to consider async_read_some - which will be just as efficient with a large buffer.

Reply:
That was definitely a mistake. Thanks for spotting this. Will fix.